### ISSUE TITLE
Adapt Unit distribution / pyro.factor from Pyro

### ISSUE THREAD

**Issue by fehiepsi** [pyro-ppl/numpyro#343]
We have some examples which uses Delta distribution to inject log_prob (hmm, neutra), so it is good to have a utility for those patterns:
+ inject a log_prob term
```
def model():
    log_prob = ...
    pyro.sample('log_prob', dist.Delta(0., log_prob), obs=0.)
```
+ convert a log_prob_fn to a distribution (without having to define a custom distribution). Currently we use the following way (which is not quite accurate): the log_prob at `x` site `dist.Normal(0, 100000)` will add a large negative *constant* to model's log_prob, which makes it harder to interprete the `log_density` output.
```
def model():
    x = pyro.sample('x', dist.Normal(0, 100000))
    log_prob = log_prob_fn(x)
    pyro.sample('log_prob', dist.Delta(0., log_prob), obs=0.)
```

**Edit:** It is better to define a custom distribution for this purpose.

A related issue is to have a utility for [transformed_pe](https://github.com/pyro-ppl/numpyro/blob/master/examples/neutra.py#L49) so users can avoid bugs related to the signs of log_det terms. It can be useful for 2 usage cases:
+ **density matching** (to sample from a known density function): When the density p(x) is known, we approximate it by a flow `flow(z)` where z ~ q(z) ~ N(0, 1) (which is what AutoContinuous does). Then instead of sampling from p(x) directly, we sample from q(z) and transform those samples using `flow`. To get those `z`, we can sample from N(0, 1) directly or using MCMC (neutra) with the objective: ``p(flow(z)).|det(flow)|``.
+ **density estimation** (to compute density from samples): When data x~p(x) is known but `p` is unknown, we use a flow `flow` to convert `p` to an easy-to-evaluate density `q=N(0,1)`, that is we are using SVI with a `None` guide and `model` is ``q(flow(x))|det(flow)|``.

In both cases, our objective is to compute base_dist's log_prob, given dist's log_prob. So it is good to have support for InverseTransform #164. In NumPyro modeling language, they are equivalent to
+ the first case
```
def model():
    # NB: we need to have a mechanism to disable automatic reparameterization
    sample('z', dist.TransformedDistribution(p_dist, flow.inv))
```
+ the second case
```
def model():
    sample('x', dist.TransformedDistribution(q_dist, flow.inv), obs=X)
```

**Comment by neerajprad** [comment]
The second issue that you mentioned above seems a bit unrelated to the issue of delta distribution. Could you create a separate issue regarding the transformed potential energy? It will be nice to follow these separately. I think the latter issue will get resolved with #348.

**Comment by fehiepsi** [comment]
Sorry, when working on #348, I think that it is better to define a custom distribution and move potential energy computation into `log_prob` method of that custom distribution. I added an edit note to the topic issue but forget to change the title. >__<
